From: Chad Monroe <chad@monroe.io>
Date: Wed, 20 Aug 2025 16:44:04 -0700
Subject: [PATCH] hostapd: pass radio idx to txpower vendor command

The mt76 txpower vendor command reads attribute 4 as the wiphy radio
index, and mt7996 rejects the command when the attribute is missing
or names no radio. hostapd sends the radio index of the hardware that
the interface uses, the same index that the EDCCA commands send.

The attribute is left out before hostapd selects that hardware and
when the wiphy reports no radios. mt7915 does not read it.

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 src/ap/ap_drv_ops.c          |    7 +------
 src/common/mtk_vendor.h      |    2 +-
 src/drivers/driver.h         |    4 ++--
 src/drivers/driver_nl80211.c |    8 +++++---
 4 files changed, 9 insertions(+), 12 deletions(-)

--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1499,15 +1499,10 @@ int hostapd_drv_get_edcca(struct hostapd
 
 int hostapd_drv_txpower_ctrl(struct hostapd_data *hapd)
 {
-	s8 link_id = -1;
-
 	if (!hapd->driver || !hapd->driver->txpower_ctrl)
 		return 0;
 
-	if (hapd->conf->mld_ap)
-		link_id = hapd->mld_link_id;
-
 	return hapd->driver->txpower_ctrl(hapd->drv_priv, hapd->iconf->lpi_psd,
 					  hapd->iconf->lpi_bcn_enhance,
-					  link_id);
+					  hostapd_drv_radio_idx(hapd));
 }
--- a/src/common/mtk_vendor.h
+++ b/src/common/mtk_vendor.h
@@ -53,7 +53,7 @@ enum mtk_vendor_attr_txpower_ctrl {
 	MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD,
 	MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX,
 	MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE,
-	MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID,
+	MTK_VENDOR_ATTR_TXPOWER_CTRL_RADIO_IDX,
 
 	/* keep last */
 	NUM_MTK_VENDOR_ATTRS_TXPOWER_CTRL,
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -97,7 +97,7 @@ txpower_ctrl_policy[NUM_MTK_VENDOR_ATTRS
 	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD] = { .type = NLA_U8 },
 	[MTK_VENDOR_ATTR_TXPOWER_CTRL_SKU_IDX] = { .type = NLA_U8 },
 	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE] = { .type = NLA_U8 },
-	[MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID] = { .type = NLA_U8 },
+	[MTK_VENDOR_ATTR_TXPOWER_CTRL_RADIO_IDX] = { .type = NLA_U8 },
 };
 
 static struct nl_sock * nl_create_handle(struct nl_cb *cb, const char *dbg)
@@ -17048,7 +17048,7 @@ static void nl80211_stop_peer_measuremen
 
 
 static int nl80211_txpower_ctrl(void *priv, u8 lpi_psd, u8 lpi_bcn_enhance,
-				u8 link_id)
+				s8 radio_idx)
 {
 	struct i802_bss *bss = priv;
 	struct wpa_driver_nl80211_data *drv = bss->drv;
@@ -17078,7 +17078,9 @@ static int nl80211_txpower_ctrl(void *pr
 	if (nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_PSD, lpi_psd) ||
 	    nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_LPI_BCN_ENHANCE,
 		       lpi_bcn_enhance) ||
-	    nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_LINK_ID, link_id))
+	    (radio_idx > -1 &&
+	     nla_put_u8(msg, MTK_VENDOR_ATTR_TXPOWER_CTRL_RADIO_IDX,
+			radio_idx)))
 		goto fail;
 
 	nla_nest_end(msg, data);
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5902,10 +5902,10 @@ struct wpa_driver_ops {
 	* @priv: Private driver interface data
 	* @lpi_psd: 1 to enable lpi psd compensate, 0 to disable
 	* @lpi_bcn_enhance: 1 to enable beacon duplicate enhancement in 6G lpi mode, 0 to disable enhancement
-	* @link_id: MLD link id. -1 if this is an non-MLD AP
+	* @radio_idx: wiphy radio index, -1 if the wiphy reports no radios
 	*/
 	int (*txpower_ctrl)(void *priv, u8 lpi_psd, u8 lpi_bcn_enhance,
-			    u8 link_id);
+			    s8 radio_idx);
 
 #ifdef CONFIG_NAN
 	/**
